README.md
# UDI Polyglot v2

[![Build Status](https://travis-ci.org/UniversalDevicesInc/polyglot-v2.svg?branch=master)](https://travis-ci.org/UniversalDevicesInc/polyglot-v2)

Polyglot is a Middleware between the [Universal Devices ISY Home Automation device](http://www.universal-devices.com/) and any number of external devices and services. 

Requirements: Network Module or ISY Portal (which includes the Network Module)

Requirements: Network Module or ISY Portal (which includes the Network Module)

Polyglot v2 is built on the [MEAN](http://mean.io/) stack which is a production ready enterprise level code stack that includes:

- MongoDB (A NoSQL database engine that stores data in JSON formated 'documents')
- Express (NodeJS HTTP API server module)
- Angular.JS (Frontend framework developed by Google)
- NodeJS (Fully asynchronous server side version of Javascript)

Using these tools, along with the [MQTT](http://mqtt.org/) messaging protocol, we have created a robust IoT powerhouse that can speak to any and every device available with open API standards using simple, straight-forward modules (NodeServers).

While most of the previous code had been written in Python, the event loop driven Node.JS was chosen for this platform due to its asynchronous nature. Python or ANY language can still be used to develop a NodeServer with the standard API interface available.

## Installation instructions

The steps are outlined directly for the Raspberry Pi, specifically the 3 or 3b models which were used to test. This has also worked on the original Raspberry Pi (Armv6) as well. A pre-built script is [here](https://raw.githubusercontent.com/UniversalDevicesInc/polyglot-v2/master/scripts/install.sh) to do all the heavy lifting for you. This install procedure was tested using a clean install of [Rasbian Stretch Lite](https://www.raspberrypi.org/downloads/raspbian/). Version **September 2017** at time of writing.

To Install Using the script (This will do EVERYTHING for you):

```bash
# Run the Install Script
wget -qO - https://raw.githubusercontent.com/UniversalDevicesInc/polyglot-v2/master/scripts/install.sh | bash -e

```

You do NOT need to proceed with theses instructions if you ran the script above. This is all done for you.

The manual steps are listed here as well.

Steps overview:

- Install MongoDB
- Install Polyglot-v2

### Prerequisites

[MongoDB](https://www.mongodb.com/)

```bash
# Update your sources
sudo apt-get update -qy

# Install the MongoDB Server / git / Python3 / Curl
sudo apt-get -qqy install mongodb-server git python3-pip python3-dev curl

# Check MongoDB Status
systemctl status mongodb

● mongodb.service - An object/document-oriented database
   Loaded: loaded (/lib/systemd/system/mongodb.service; enabled)
   Active: active (running) since Thu 2017-07-13 03:35:41 UTC; 25min ago
     Docs: man:mongod(1)
 Main PID: 431 (mongod)
   CGroup: /system.slice/mongodb.service
           └─431 /usr/bin/mongod --config /etc/mongodb.conf
```

### Polyglot

Now the good stuff. Let's install Polyglot! Login as the user that is going to run Polyglot (pi or whatever you choose). Polyglot does NOT need to be run as root! (PLEASE DO NOT RUN POLYGLOT AS ROOT)

```bash
# Create a polyglot directory
mkdir polyglot

# Enter polyglot directory
cd polyglot

# Get the Polyglot binary. This example is for armv7 Raspberry Pi 3/3b. (armv6 is RPi 1/Nano and x64 is most x86 linux boxes)
# If you don't know which one to get run the command 'lscpu' and look at the top line. 'Architecture'
wget -q https://s3.amazonaws.com/polyglotv2/binaries/polyglot-v2-linux-armv7.tar.gz

# Extract the archive
tar -zxf polyglot-v2-linux-armv7.tar.gz

# Run Polyglot normally:
./polyglot-v2-linux-armv7

# OR

# Run Polyglot with debug logging turned on:
NODE_ENV=development ./polyglot-v2-linux-armv7
```

To start Polyglot on Boot do the following:

```bash
# Get base systemd script from the GitHub repository
wget -q https://raw.githubusercontent.com/UniversalDevicesInc/polyglot-v2/master/scripts/polyglot-v2.service

# Edit it for your system
nano polyglot-v2.service

# Modify the following sections in the polyglot-v2.service file
# Replace <PWD> with the current directory that the polyglot binary resides in.
# Replace <USER> with the user you want polyglot to run as
# Replace <FILE> with the FULL path to the polyglot binary. e.g. /home/pi/polyglot/polyglot-v2-linux-armv7

# One the above changes are done to the polyglot-v2.service file, move it to the systemd unit folder
sudo mv polyglot-v2.service /lib/systemd/system/

# Add the service then start polyglot-v2
sudo systemctl enable polyglot-v2
sudo systemctl start polyglot-v2
```

Verify that Polyglot is running

```bash
# Check Polyglot status
sudo systemctl status polyglot-v2.service

● polyglot-v2.service - polyglot-v2
   Loaded: loaded (/lib/systemd/system/polyglot-v2.service; enabled)
   Active: active (running) since Thu 2017-07-13 04:38:53 UTC; 10min ago
 Main PID: 1647 (node)
   CGroup: /system.slice/polyglot-v2.service
           └─1647 node /usr/local/bin/polyglot-v2

Jul 13 04:38:53 raspberrypi systemd[1]: Started polyglot-v2.
```

Default username and password are both `admin`

Logs are located at `~/.polyglot/log/debug.log` or you can use the frontend web interface to watch them in real-time.

```bash
# Watch the logs real-time
tail -f ~/.polyglot/log/debug.log

7/13/2017, 4:24:15 AM - info: Starting Polyglot version 2.0
7/13/2017, 4:24:15 AM - info: MongoDB: Connected
7/13/2017, 4:24:16 AM - info: MQTT Services Started.
7/13/2017, 4:24:16 AM - info: Secure Server started on port 3000
7/13/2017, 4:24:16 AM - info: ISY: Got Version 5.0.10
7/13/2017, 4:24:16 AM - info: MQTT: Subscribe Successful udi/polyglot/connections/# QoS: 0
7/13/2017, 4:24:38 AM - info: Successful login by admin
7/13/2017, 4:24:39 AM - info: MQTT: Frontend Websockets interface  Connected.
```

#### Enable debug logging for Polyglot

```bash
# Edit the ~/.polyglot/.env file
nano ~/.polyglot/.env

# Add or modify the NODE_ENV variable
NODE_ENV=development

```

#### Configuration Overrides for Polyglot

These configuration overrides are available in the ~/.polyglot/.env file. It is not required so you may have to create it. Most of them can be updated via the web interface and are all saved to the database, so these are typically not required unless you have a specific reason. These WILL override existing database settings, even if you change them and save in the frontend. Remember these are OVERRIDES.

```bash
# Overrides the IP address Polyglot listens on the local machine.
BIND_IP='192.168.1.2'

# Overrides the IP address Polyglot sends to ISY to accept commands
HOST_IP='192.168.1.2'

# Overrides the Port that Polyglot listens on for its frontend interface.
HOST_PORT='3000'

# Overrides the default mode of HTTPS
USE_HTTPS=false

# Username used to login to the ISY.
ISY_USERNAME='admin'

# Password used to login to the ISY. Careful this is clear text, I wouldn't recommend setting this here.
ISY_PASSWORD='password'

# ISY IP address. This is automatically discovered on the initial run of Polyglot if you are on the same network. If you have multiple you can update it on the settings page of the frontend, or override it here.
ISY_HOST='192.168.1.10'

# ISY Port
ISY_PORT='80'

# ISY HTTPS: True/False This isn't fully tested so beware HTTPS at the moment.
ISY_HTTPS=false

# MQTT Host is the IP address of the host running a MQTT server. This is built in to Polyglot so you won't need this unless you'd prefer an external MQTT server.
MQTT_HOST='127.0.0.1'

# MQTT Port is the port used to connect to the MQTT server. Default is 1883
MQTT_PORT='1883'

# URI to access MongoDB. You might need this if you have an off-box Mongo instance.
MONGO_URI='mongodb://localhost:27017/'

# To enable debug logging set the NODE_ENV override to 'development'
NODE_ENV='development'

# To enable Custom SSL Certificates
CUSTOM_SSL=true

# Enable Beta Software Updates
USE_BETA=true
```

#### Custom SSL Certificates for Polyglot

If you want to use your own SSL certificates instead of the self-signed certificates that Polyglot auto generates for you, then place the certificates in the ~/.polyglot/ssl/custom/ folder. There should be three files present as outlined below. The custom.key file is read then encrypted in the database, the other two are simply read and stored since they are public certs. If the CUSTOM_SSL .env override is set to true, polyglot attempts to read the following files:

- custom.key: this is your private key file
- custom.crt: this is your public cert file
- custom.ca: this file can contain multiple ca certificates, otherwise known as the trust chain, this file will have any intermediate and root certificates

Polyglot will attempt to read these files on startup EVERY TIME that CUSTOM_SSL is enabled. If you allow Polyglot to read the files once, you can then delete the custom files and Polyglot will continue to use the saved values from the database. DO NOT DELETE THE ~/.polyglot/ssl/ client/polyglot certificates, these are used for MQTT encryption and authentication. Polyglot will spit them back out on startup even if you accidentally delete them.

#### SSL Error 'ee key too small'

I've had several linux users run into the error 'ee key too small' when trying to start Polyglot.

`sudo nano /etc/ssl/openssl.cnf`

Change the line:

```bash
CipherString = DEFAULT@SECLEVEL=2
to
CipherString = DEFAULT@SECLEVEL=1
```

### Backup and Restore

As of version 2.2.4+ you can now backup and restore your NodeServers and Nodes on to a different installation of Polyglot.
Click on Settings > Settings and you will find a "Backup and Restore" section.

Backup will download a .bin file to your machine that is an encrypted copy of all installed NodeServers and Nodes. This does NOT include any other settings like ISY connection settings (on purpose). So if you intend to migrate to a new installation of Polyglot here are the steps:

- Download backup from existing installation
- Install and configure ISY connection settings of new installation of Polyglot
- Restore backup

To Restore the backup, click the "Select File" button on the "Backup and Restore" section of the settings menu. Once you select the file, you will see that same button change to "Restore". Once you hover over that button you will see the selected filename that you chose to restore. Click the "Restore" button to initiate the restore process. This does the following:

- Decrypts and validates the file
- Individiually iterates through the backed up NodeServers doin the following:
  - Verifies a NodeServer is not currently running and active on that slot (shuts it down for overwrite if there is)
  - Overwrites or Creates the NodeServer database entry
  - Overwrites or Creates any Nodes in the database associate with that NodeServer
  - Overwrites or Creates the slot configuration on the ISY for that NodeServer (this includes updating the existing ISY configuration for that NodeServer to the new IP address of Polyglot, **WITHOUT** removing nodes and re-creating them, thus saving your existing programs.)
  - Downloads or Pulls the latest code for that NodeServer automatically
  - Installs the profile to the ISY from that freshly downloaded NodeServer code

Once all NodeServers have had the previous steps completed, it will automatically shut down Polyglot. If you have systemctl set up properly (as you should), Polyglot will automatically restart and everything should be working appropriately with no further interaction necessary.

### Development Documentation

First check the [Wiki](https://github.com/UniversalDevicesInc/polyglot-v2/wiki/Creating-a-NodeServer) for how to build a NodeServer.

It is pre-built in the docs folder [Found Here](https://github.com/UniversalDevicesInc/polyglot-v2) on GitHub in HTML format if you clone the repository or available in the
[Online Documentation Found Here](https://doclets.io/Einstein42/udi-polyglotv2/master/overview).

Python3 interface API for building NodeServers is [here](https://github.com/UniversalDevicesInc/polyglot-v2-python-interface).

NodeJS interface API for building NodeServers is [here](https://github.com/UniversalDevicesInc/pgc-nodejs-interface).


CHANGELOG.md
# __Changelog for Polyglot v2__

## Version 2.2.11

* Update Aedes to 0.44.2
* Fix to create link to default debug.log name
* Fix Download Log Package to include debug.log data instead of soft link.

## Version 2.2.10

* Updated Aedes to 0.41.0
* Change Logging date/time format to match Python interface
* Log can be 100m, save 14 days with compression of rotated log each day

## Version 2.2.9

* ispolisy Frontend API added, modified Polisy configuration page
* Polisy Frontend formatting bug fixes
* Polisy Frontend IPv6 additions
* Fixed Polyglot/Polisy real time detection mechanism
* Fixed Polisy detect ISY hang
* Added additional logging for ISY/Polisy detection routine

## Version 2.2.8

* Polisy Frontend added to menus and new page

## Version 2.2.7

* Zipped backup file and added persist_folder in server.json to capture nodeserver files as needed
* Fixed remove nodeserver bug, where node database entries could persist after delete
* Fixed log rotation bug. Logs will rotate at 1mb in size and keep 5
* Added Polisy System Configuration GUI

## Version 2.2.6

* Fixed shutdown when large amount of nodeservers were installed
* Fixed override of port on Polisy
* Fixed Frontend redundant MQTT connection on reboot
* Modified Log page format

## Version 2.2.5

* FreeBSD upgrade
* FreeBSD Polisy upgrade (pkg)
* Configurable ListenPort on the Settings page
* Restore fix bug on HomeDir

## Version 2.2.4

* Updated versions of all packages to latest
* Winston 2 > 3
* MQTT 2 > 3
* Fixed Encryption Deprecation Warning
* Fixed Buffer Deprecation Warning
* Fixed Mongoose/MongoDB deprecations
* Backup/Restore functionality through settings menu for nodeservers/nodes
* Automatically update ip adderss on ISY for nodeservers restored to new devices
* Fully encrypted backup files
* Precursor to 2.3.0 for Polisy

## Version 2.2.3

* Angular 8
* Added automated patch and release for versioning
* Dockerfile update

## Version 2.2.2

* Fixed Logging
* Hints fixed
* Config fix refresh

## Version 2.2.1

* Miscellaneous bug fixes

## Version 2.2.0

* NodeJS 9.1.2
* Frontend changes to support log package and log downloading
* Frontend organization cleanup
* Added Polyglot restart button
* Cleaned up log windows with scrolling
* Fixed several nodeservers not showing acurate status

## Version 2.1.2

* Moved addNode/updateNode logic into Polyglot removed from polyinterfaces
* Added update script to run on all future upgrades to automatically upgrade polyinterfaces
* Added change nodeDef call if nodeDefId is changed for a node
* Added change API for ns to change nodeDefId if necessary
* sendResult bug fix where sometimes node.start() wasn't called
* Documentation updates

## Version 2.1.1

* Upgraded to Angular 5 for Frontend interface
* Migrated from Paho to MQTT.JS for Frontend MQTT Client
* Fixed Angular 5 Modal
* Added Uptime/Started At Frontend messages
* Fixed MQTT inconsistencies with Aedes
* Completely re-wrote the db calls
* Moved to ISY Keep-alive for increased speed
* Increased overall speed roughly 3000% on adds/driver updates
* Dropped CPU/mem usage by roughly 60%
* Drivers automatically sync to ISY on node add for 'initialization'
* Drivers automatically sync to ISY on startup from previously save values
* Restructured database, no breaking changes at the moment, eventually moving to multi-add interface
* Allow HTTP non-secure configuration option
* Specify IP to help with Docker builds

## Version 2.1.0

* Fixed graceful shutdowns and error catches to properly (and cleanly) shut down.
* Binaries moved to UDI AWS S3 bucket. Removed from github completely
* Fixed update node not pulling the new version correctly
* Fixed uninstall nodeserver failing
* Restructured most of the project in prep for unit testing
* Added check for ISY firmware > 5.0.0
* Changed to read profile directly instead of profile.zip
* Moved from Mosca MQTT Server to Aedes
* Major update to startup/shutdown and data flows
* Added NodeServer self-service addition button
* Added documentation link to github repo for each NodeServer

## Version 2.0.36

* Binaries moved to S3 (will remove from git completely in 2.1.0)
* Dynamic lamdba Store and added NodeServer version support (thanks xKing)
* Fixed per nodeserver queue for stability and rate limits
* Re-wrote majority of functions into es2017 async/await structures
* Re-wrote startup functions
* Re-structured dependencies
* Overhaul of all intrinsic mqtt functions to be more efficient
* Changed store layout
* Added customData for nodeserver storage
* Removed auto-reboot functions as they are no longer necessary

## Version 2.0.35

* Moved to UDI github Repo
* Completely re-wrote the addnode functions to not be terrible, moved to ES7+ async await
* Added NodeServer customdata API
* Implemented inQ's per NodeServer to throttle inbound traffic
* Implemented outQ to ISY for concurrency to help connection stability
* Started code cleanup and unit testing
* Restores deleted ISY nodes on NodeServer restart
* Handles retries to ISY for connection errors

## Version 2.0.32 - 34

* Worked on duplicate node bug

## Version 2.0.31

* Added custom certificate ability, see README
* Fixed addnode bug where duplicates were possible
* Fixed removenode callback bug
* Multiple primaries can now be added (requires polyinterface 2.0.6+)
* Dynamic updates to drivers by allowing overwrite adds to nodes

## Version 2.0.30

* Fixed spacing on NodeServer Store page
* Skip Auto-Detection of ISY if ISY_HOST is defined in .env
* Added prompt to update profile and reboot on NodeServer git update
* Cleaned up some frontend message parsing

## Version 2.0.29

* Fixed customParams deletion of last key from frontend
* Verified port override works in .env
* Added better connection handling to frontend
* Prominent disconnect notice from backend
* Better reconnection and recovery mechanism for Frontend
* Added shortPoll/longPoll configuration items

modules/mqtt.js
const logger = require('./logger')
const mqtt = require('mqtt')
const config = require('../config/config')

module.exports = {
		Client: null,
		clientId: null,

		startService(callback) {
			if (this.Client) { if (callback) return callback() }
			var options = {
				keepalive: 60,
				clean: true,
				clientId: 'polyglot',
				reconnectPeriod: 5000,
				will: {retain: true}
			}
			if(!process.env.MQTT_HOST || !process.env.MQTT_PORT) {
				logger.error('MQTT Address or Port not set... Check your Settings.')
				return
			}

			this.clientId = 'polyglot'
			options['clientId'] = this.clientId
			/*
			Set the Last Will and Testament for the Nodeserver. Allows NodeServers to be notified
			in case	of catastrophic failure. This message is automatically stored and sent by the
			broker if ungraceful disconnection occurs or the keepalive is exceeded.
			*/
			options['will']['topic'] = 'udi/polyglot/connections/polyglot'
			options['will']['payload'] = new Buffer(JSON.stringify({node: this.clientId, 'connected': false}))
			this.Client = mqtt.connect('mqtt://'+ process.env.MQTT_HOST + ':' + process.env.MQTT_PORT, options)

			this.Client.on('connect', () => {
				this.addSubscriptions()
			})

			this.Client.on('message', (topic, payload) => {
				try {
					payload = JSON.parse(payload.toString())
				} catch (e) {
					logger.error('MQTT: Badly formatted JSON input received. ' + e)
				}
				this.parse(topic, payload)
			})

			this.Client.on('reconnect', () => {
				config.mqttConnected = false
				logger.info('MQTT attempting reconnection to broker...')
			})

			this.Client.on('error', (err) => {
				logger.error('MQTT recieved error: ' + err.toString())
			})

      logger.info('MQTT Services Started.')
			if (callback) callback()
		},

		addSubscriptions() {
			this.Client.subscribe('udi/polyglot/connections/#', (err, granted) => {
				if (err) {
					logger.error('Error: ' + err.toString())
					return
				}
				config.mqttConnected = true
				logger.info('MQTT: Subscribe Successful ' + granted[0]['topic'] + " QoS: " + granted[0]['qos'])
				this.publish('udi/polyglot/connections/polyglot', {node: this.clientId, 'connected': true}, { retain: true })
			})
			this.Client.subscribe('udi/polyglot/frontend/settings')
			this.Client.subscribe('udi/polyglot/frontend/nodeservers')
			config.nodeServers.forEach((nodeServer) => {
				this.Client.subscribe('udi/polyglot/ns/' + nodeServer.profileNum, (err, granted) => {
					if (err) {
						logger.error('Error: ' + err.toString())
						return
					}
					logger.info('MQTT: Subscribe Successful ' + granted[0]['topic'] + " QoS: " + granted[0]['qos'])
				})
			})
		},

		addSubscription(profileNum) {
			this.Client.subscribe('udi/polyglot/ns/' + profileNum, (err, granted) => {
				if (err) {
					logger.error('Error: ' + err.toString())
					return
				}
				logger.info('MQTT: Subscribe Successful ' + granted[0]['topic'] + " QoS: " + granted[0]['qos'])
			})
		},

		delSubscription(profileNum) {
			this.Client.unsubscribe('udi/polyglot/ns/' + profileNum, (err, granted) => {
				if (err) {
					logger.error('Error: ' + err.toString())
					return
				}
				logger.info(`MQTT: Unsubscribed Successfully from NodeServer ${profileNum}`)
			})
		},


		stopService(callback) {
			if(this.Client){
					this.publish('udi/polyglot/connections/polyglot', {node: this.clientId, 'connected': false}, { retain: true })
					logger.info('MQTT Services Stopping Gracefully.')
					this.Client.end(true, () => {
						this.Client = null
						if(callback) { callback() }
					})
			} else {
					if(callback) { callback() }
			}
		},

		makeResponse(topic, command, message) {
			if (topic === 'connections' || topic === 'udi/polyglot/connections/polyglot') {
				topic = 'udi/polyglot/connections/polyglot'
			} else {
				topic = 'udi/polyglot/ns/' + topic
			}
      try {
        var response = {'node': 'polyglot'}
        response[command] = message
      } catch (e) {
				console.log(e)
         var response = {
          'node': 'polyglot',
          'data': {
            'error': e
          }
        }
      }
			this.publish(topic, response)
    },

		publish(topic, message, options, callback) {
			message = JSON.stringify(message)
			this.Client.publish(topic, message, options, callback)
		},

		parse(topic, message) {
			var parse = false
			if (message.node === 'polyglot') { return }
			if (topic === 'udi/polyglot/frontend/settings') {
				require('../models/settings').updateSettings(message)
				return
			}
			if (topic === 'udi/polyglot/frontend/nodeservers') {
				require('../models/nodeserver').parseFrontend(message)
				return
			}
			logger.debug('MQTT Message: ' + topic + ": " + JSON.stringify(message))
			if (topic.substring(0,25) === 'udi/polyglot/connections/') {
				if (message.node.substring(0,18) === 'polyglot_frontend-') {
					logger.info('MQTT: Frontend Websockets interface ' + (message.connected ? ' Connected.' : ' Disconnected.'))
				} else if (message.node === config.nodeServers[message.node].profileNum) {
						try {
							config.nodeServers[message.node].checkCommand(message)
						} catch (e) {
							logger.error(`MQTT CheckCommand Error: ${e}`)
						}
				}
			} else if ((topic.substring(0,16) === 'udi/polyglot/ns/') && (topic.slice(-1) === message.node)) {
				if (message.node === config.nodeServers[message.node].profileNum) {
					try {
						config.nodeServers[message.node].checkCommand(message)
					} catch (e) {
						logger.error(`MQTT CheckCommand Error: ${e}`)
					}
				}
			} else {
				logger.debug('MQTT: Did not match any parse filters. Ignoring. This usually means ' +
				 'the node value is incorrect. Make sure it matches an active NodeServer and you are publishing to the correct topic. ' + message)
			}
		}
}
